apps/prairielearn/src/lib/course.ts

```diff
@@ -162,6 +162,10 @@ export async function pullAndUpdateCourse({
 
         job.info('Sync git repository to database');
         const syncResult = await syncDiskToSqlWithLock(courseId, path, job);
+        if (syncResult.hardFail) {
+          job.fail('Sync completely failed due to invalid question sharing edit.');
+          return;
+        }
```

Comment:
After reviewing this, I have a question: should we `git reset` to the previous commit hash on failure here? I'm thinking about the case where someone clicks the "Delete question" button in the UI. If we commit/push the deletion, they have no good way to get the question back without mucking around in the Git history. Obviously, we should disable that button in this case, but generally speaking it seems like it'd be most robust if we reset to a known good state.

This would of course change where we want this `job.fail()` call to go. If we want to reset, the current strategy would be fine. I'd be inclined to make `hardFail` more specific though, especially since the error message is _very_ specific to question sharing.

Reply:
Actually, syncDiskToSqlWithLock isn't always called from within a job, it doesn't have access to the job, so that doesn't really work (at least with the way things are set up now).


From what you have said, `git reset` seems like a good idea, but I know very little about how PrairieLearn interacts with git, so my opinion is probably not worth much. Should we ask Matt's opinion?


You have a good point about `hardFail` not being a good name. Should I do `sharingSyncError` instead?

Reply:
Good point about this not always being called in the context of a job. I saw _one_ call site where `syncDiskToSqlWithLock` is called with a job, but that's actually something following the `Logger` interface.  

Yeah, can you queue this up for discussion at the dev meeting this week? If you'd like to resolve this sooner, you can start a discussion on Slack.